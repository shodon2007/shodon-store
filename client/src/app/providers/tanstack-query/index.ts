import ReactQuery from './ui/tanstack-query'

export { ReactQuery }
