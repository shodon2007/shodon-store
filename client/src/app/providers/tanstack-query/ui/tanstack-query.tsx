import {
    QueryClient,
    QueryClientProvider,
} from '@tanstack/react-query';
import { FC } from 'react';

const queryClient = new QueryClient();

interface ReactQueryProps {
    children: React.ReactNode,
}

const ReactQuery: FC<ReactQueryProps> = ({ children }) => {
    return (<QueryClientProvider client={queryClient} >
        {children}
    </QueryClientProvider>
    )
}

export default ReactQuery;