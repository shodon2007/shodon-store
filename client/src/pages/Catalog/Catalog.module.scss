.catalog {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    gap: 20px;
}

.item {
    background-color: white;
    border: 3px solid white;
    cursor: pointer;
    display: flex;
    justify-content: space-between;
    flex-direction: column;
    text-align: center;
    min-height: 200px;
    gap: 20px;
    padding: 10px 30px;
    font-size: 16px;
    color: #1a2733;
    text-decoration: none;
    transition: all 0.3s ease;

    &:hover,
    &:focus {
        border: 3px solid #0385ff;
        color: #0385ff;
    }
}

.loading {
    min-height: 200px;
    background-color: white;
    display: flex;
    align-items: center;
    justify-content: center;

    & > img {
        animation: loadingAnimation 3s linear infinite;
        width: 50px;
        height: 50px;
    }
}

.img {
    width: 100%;
    background-color: white;
    padding: 5px;
    transition: all 0.3s ease;
    flex: 1 1 auto;
    display: flex;
    align-items: center;
    justify-content: center;
}

.img > img {
    max-width: 100%;
    max-height: fit-content;
}

@media (max-width: 720px) {
    .catalog {
        grid-template-columns: repeat(2, 1fr);
    }

    .item {
        padding: 10px 20px;
    }
}

@media (max-width: 360px) {
    .item {
        padding: 5px 10px;
    }
}

@keyframes loadingAnimation {
    0% {
        transform: rotate(0deg);
    }

    100% {
        transform: rotate(360deg);
    }
}
