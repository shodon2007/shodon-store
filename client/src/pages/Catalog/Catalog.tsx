import { FC } from "react";
import classes from "./Catalog.module.scss";
import { useQuery } from "@tanstack/react-query";
import catalogService from "../../services/catalog.service";
import { URL } from "../../global";
import { Link } from "react-router-dom";
import Loading from "../../static/loading.webp";

const Catalog: FC = () => {
    const { data, isLoading } = useQuery({
        queryKey: ["catalog"],
        queryFn: () => catalogService.getCatalog(),
    });

    if (isLoading) {
        return <Sceleton />;
    }

    if (!data) {
        return <div>ашибка, ну вообще врятли эта будет</div>;
    }

    return (
        <div className={classes.catalog}>
            {data.map((catalog, i) => {
                return (
                    <Link to={catalog.name} key={i} className={classes.item}>
                        <div className={classes.img}>
                            <img src={`${URL}/${catalog.img}`} alt="img" />
                        </div>
                        <span>{catalog.name_ru}</span>
                    </Link>
                );
            })}
        </div>
    );
};

const Sceleton = () => {
    const sceletons = [];

    for (let i = 0; i < 6; i++) {
        sceletons.push(
            <div className={classes.loading} key={i}>
                <img src={Loading} alt="loading" />
            </div>
        );
    }
    return <div className={classes.catalog}>{sceletons.map((i) => i)}</div>;
};

export default Catalog;
