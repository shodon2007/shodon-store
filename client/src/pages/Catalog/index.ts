import CatalogPage from "./ui/Catalog";

export {
    CatalogPage,
}