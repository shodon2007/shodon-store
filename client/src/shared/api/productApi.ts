import { FilterType } from './filterApi'
import ApiGeneral from './generalApi'

export interface Attribute {
  id?: number
  title: string
  description: string
}

export interface CatalogItem {
  name_ru: string,
  img: string,
  id: number,
  name: string,
}

export interface Product extends CatalogItem {
  price: number
  brand_id: number
  rate: string
  reviews: string
  type_id: number
  attributes: Attribute[]
}

class ProductApi extends ApiGeneral {
  async getAllProducts (filters: FilterType, type: string) {
    const res = await this.api.get<Product[]>(this.urls.products, {
      params: {
        filters,
        type,
      }
    });
    console.log(res);
    return res;
  }
}

export default new ProductApi()
