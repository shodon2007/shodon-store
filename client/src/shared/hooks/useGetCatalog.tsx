import { useQuery } from "@tanstack/react-query"
import catalogApi from "src/shared/api/catalogApi";

const useGetCatalog = () => {
    const query = useQuery({
        queryKey: ['catalog'],
        queryFn: () => {
            return catalogApi.getCatalog().then(res => res.data);
        },
    });

    return query;
}

export {
    useGetCatalog
}