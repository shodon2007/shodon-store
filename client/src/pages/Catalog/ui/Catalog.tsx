import { FC } from 'react';

import { useGetCatalog } from 'src/shared/lib/useGetCatalog/useGetCatalog';

import cls from './Catalog.module.scss';
import { Link } from 'react-router-dom';
import { serverUrl } from 'src/shared/api/generalApi';

const CatalogPage: FC = () => {
	const { data: catalog, isLoading, error, isError } = useGetCatalog();

	if (isLoading) {
		return (
			<span data-testid='loading-page'>Загрузка каталога, подождите...</span>
		);
	}

	if (isError) {
		return error.message;
	}

	return (
		<div className={cls.catalog} data-testid='catalog-page'>
			{catalog.map(elem => {
				return (
					<Link
						to={`/${elem.name}`}
						className={cls.item}
						key={elem.id}
						data-testid='catalog-link'
					>
						<div className={cls.img}>
							<img src={`${serverUrl}/img${elem.img}`} alt={elem.name} />
						</div>
						<span className={cls.name}>{elem.name_ru}</span>
					</Link>
				);
			})}
		</div>
	);
};

export default CatalogPage;
