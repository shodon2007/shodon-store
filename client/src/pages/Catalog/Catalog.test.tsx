import { render, screen } from "@testing-library/react"
import Catalog from "./Catalog"
import { ReactQuery } from "src/app/providers/tanstack-query"
import { useGetCatalog } from "src/shared/hooks/useGetCatalog"
import { CatalogItem, Product } from "src/shared/api/productApi";
import axios, { AxiosHeaders, AxiosResponse } from "axios";
import { UseQueryResult } from "@tanstack/react-query";

jest.mock('axios');

describe('Testing Catalog.test.tsx', () => {
    let resp: AxiosResponse<CatalogItem[], any> | {} = {
    }

    beforeEach(() => {
        resp = {
            config: {
                headers: new AxiosHeaders(),
            },
            data: [
                {
                    name_ru: 'Телефоны',
                    img: 'https://google.com',
                    id: 1,
                    name: 'phones',
                }
            ],
            status: 200,
            statusText: 'hello'
        }
    })

    test('first testging', async () => {
        (axios.get as jest.Mock).mockReturnValue(resp);
        render(<ReactQuery><Catalog /></ReactQuery>)
        const items = await screen.findAllByTestId('catalog-link');
        expect(items.length).toBe(1)
    })
})