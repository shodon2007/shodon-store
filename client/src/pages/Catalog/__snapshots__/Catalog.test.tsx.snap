// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Catalog.test.tsx Testing Catalog links 1`] = `
<div
  class="catalog"
  data-testid="catalog-page"
>
  <a
    class="item"
    data-testid="catalog-link"
    href="/hello"
  >
    <div
      class="img"
    >
      <img
        alt="hello"
        src="https://store-server.shodon.ru/imghedfsdkf"
      />
    </div>
    <span
      class="name"
    >
      прив
    </span>
  </a>
  <a
    class="item"
    data-testid="catalog-link"
    href="/Телефончики"
  >
    <div
      class="img"
    >
      <img
        alt="Телефончики"
        src="https://store-server.shodon.ru/imghttps://hedfsdkf.png"
      />
    </div>
    <span
      class="name"
    >
      Phones
    </span>
  </a>
</div>
`;

exports[`Testing Catalog.test.tsx first testging 1`] = `
<div
  class="catalog"
  data-testid="catalog-page"
>
  <a
    class="item"
    data-testid="catalog-link"
    href="/hello"
  >
    <div
      class="img"
    >
      <img
        alt="hello"
        src="https://store-server.shodon.ru/imghedfsdkf"
      />
    </div>
    <span
      class="name"
    >
      прив
    </span>
  </a>
  <a
    class="item"
    data-testid="catalog-link"
    href="/Телефончики"
  >
    <div
      class="img"
    >
      <img
        alt="Телефончики"
        src="https://store-server.shodon.ru/imghttps://hedfsdkf.png"
      />
    </div>
    <span
      class="name"
    >
      Phones
    </span>
  </a>
</div>
`;

exports[`Testing Catalog.test.tsx testing loading 1`] = `
<span
  data-testid="loading-page"
>
  Загрузка каталога, подождите...
</span>
`;
